Lazily realize IOperation.ConstantValue

333fred opened the pull request:
Today, we realize ConstantValues of nodes as soon as an IOperation is created. This means that for things like constant string addition, we immediately use O(n^2) memory when an IOperation tree is fully realized. Internally the compiler uses an approach to only realize the final string when necessary, and this ports IOperation to use that representation under the hood for strings. When an IOperation.ConstantValue that is a string is realized, we take a WeakReference to that string. If a subsequent request comes in and the reference has not yet expired, we return that same string to avoid duplication, but avoid rooting the object if one user requests the constant value and then immediately gets rid of it. Fixes https://github.com/dotnet/roslyn/issues/43019.

333fred commented:
In validating this PR, I ran locally with the IOperation test hook enabled and confirmed that there were no new failures from this change.

333fred commented:
@dotnet/roslyn-compiler @AlekseyTs this PR is ready for review. /cc @mavasani @CyrusNajmabadi.

AlekseyTs commented on src/Compilers/Core/Portable/Operations/Operation.cs:
```diff
@@ -19,31 +20,33 @@ namespace Microsoft.CodeAnalysis
+            Debug.Assert(constantValue is object);
             // Constant value cannot be "null" for non-nullable value type operations.
-            Debug.Assert(type?.IsValueType != true || ITypeSymbolHelpers.IsNullableType(type) || !constantValue.HasValue || constantValue.Value != null);
+            Debug.Assert(type?.IsValueType != true || ITypeSymbolHelpers.IsNullableType(type) || !constantValue.Value.HasValue || constantValue.Value.Value != null);
```
>Value.HasValue [](start = 112, length = 14)

Consider adding an API to the OperationConstantValue to answer if ther is a value and whether the value is null without realizing the value.

333fred replied:
Removed `OperationConstantValue` entirely.

---
In reply to: [444569843](https://github.com/dotnet/roslyn/pull/45127#discussion_r444569843) [](ancestors = 444569843)

AlekseyTs commented on src/Compilers/Core/Portable/Operations/Operation.cs:
```diff
@@ -119,28 +122,29 @@ public string Language
             get => Syntax.Language;
         }
 
+        internal OperationConstantValue OperationConstantValue { get; }
```
>internal OperationConstantValue OperationConstantValue { get; } [](start = 8, length = 63)

Consider adding an empty line after this line.

AlekseyTs commented on src/Compilers/Core/Portable/Operations/OperationConstantValue.cs:
```diff
@@ -0,0 +1,95 @@
+            /// value if no users are still using it.
+            /// </summary>
+            private WeakReference<string>? _constantValueReference;
+            private readonly ConstantValue _originalConstantValue;
```
>private readonly ConstantValue _originalConstantValue; [](start = 12, length = 54)

Should we push the weak reference into the ConstantValue implementation that needs to build the string value?

333fred replied:
Removed this class entirely.

---
In reply to: [444572513](https://github.com/dotnet/roslyn/pull/45127#discussion_r444572513) [](ancestors = 444572513)

AlekseyTs commented:
>             OperationConstantValue constantValue = operand.ConstantValue.HasValue

I think we should try to avoid realizing the value just to check if it is null.

---
Refers to: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:3093 in 022b6e6. [](commit_id = 022b6e65e851ee0a9380d80948b18587e2bf4778, deletion_comment = False)

AlekseyTs commented on src/Compilers/Core/Portable/Operations/OperationConstantValue.cs:
```diff
@@ -0,0 +1,95 @@
+    /// <summary>
+    /// Root type for operation constant values.
+    /// </summary>
+    internal abstract class OperationConstantValue
```
>OperationConstantValue [](start = 28, length = 22)

It feels like creating a wrapper is unnecessary, simply holding onto ConstantValue instance should be sufficient.

AlekseyTs commented on src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs:
```diff
@@ -1983,7 +1983,7 @@ internal override IConvertibleConversion ClassifyConvertibleConversion(IOperatio
             {
                 if (source.ConstantValue.HasValue && source.ConstantValue.Value is null && destination.IsReferenceType)
```
>source.ConstantValue.HasValue && source.ConstantValue.Value is null [](start = 20, length = 67)

I think we should avoid realizing a value for the purpose of this check.

AlekseyTs replied:
Please review references to ConstantValue in our codebase to locate all other similar checks that could be optimized.

---
In reply to: [444576043](https://github.com/dotnet/roslyn/pull/45127#discussion_r444576043) [](ancestors = 444576043)

333fred replied:
> Please review references to ConstantValue in our codebase to locate all other similar checks that could be optimized.

Done. I've left the test calls alone, as they're testing the codepath, but I've fixed the ones here I saw in the mainline projects (mainly here and in CFGBuilder)

---
In reply to: [444577989](https://github.com/dotnet/roslyn/pull/45127#discussion_r444577989) [](ancestors = 444577989,444576043)

AlekseyTs commented on src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs:
```diff
@@ -1994,7 +1994,7 @@ internal override IConvertibleConversion ClassifyConvertibleConversion(IOperatio
 
             if (result.IsReference && source.ConstantValue.HasValue && source.ConstantValue.Value is null)
```
>source.ConstantValue.HasValue && source.ConstantValue.Value is null [](start = 38, length = 67)

Same here

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Operations/VisualBasicOperationFactory_Methods.vb:
```diff
@@ -11,8 +11,8 @@ Imports Microsoft.CodeAnalysis.VisualBasic.Syntax
     Partial Friend NotInheritable Class VisualBasicOperationFactory
-        Private Shared Function ConvertToOptional(value As ConstantValue) As [Optional](Of Object)
-            Return If(value Is Nothing OrElse value.IsBad, New [Optional](Of Object)(), New [Optional](Of Object)(value.Value))
+        Private Shared Function ConvertToOptional(value As ConstantValue) As OperationConstantValue
```
>ConvertToOptional [](start = 32, length = 17)

I think this helper was renamed on C# side. BTW, can we share is between languages?

333fred replied:
Removed the methods entirely.

---
In reply to: [444579977](https://github.com/dotnet/roslyn/pull/45127#discussion_r444579977) [](ancestors = 444579977)

AlekseyTs commented:
Done with review pass (iteration 1)

333fred commented:
@AlekseyTs addressed feedback.

---
In reply to: [648514193](https://github.com/dotnet/roslyn/pull/45127#issuecomment-648514193) [](ancestors = 648514193)

AlekseyTs commented on src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs:
```diff
@@ -1992,9 +1993,11 @@ internal override IConvertibleConversion ClassifyConvertibleConversion(IOperatio
             Conversion result = ClassifyConversion(sourceType, destination);
 
-            if (result.IsReference && source.ConstantValue.HasValue && source.ConstantValue.Value is null)
             {
```
>{ [](start = 12, length = 1)

Addition of a block looks unintentional.

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Compilation/VisualBasicCompilation.vb:
```diff
@@ -1816,9 +1816,10 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
             If sourceType Is Nothing Then
-                If source.ConstantValue.HasValue AndAlso source.ConstantValue.Value Is Nothing AndAlso destination.IsReferenceType Then
-                    constantValue = source.ConstantValue
+                If sourceConstantValue?.IsNothing = True AndAlso destination.IsReferenceType Then
```
>sourceConstantValue?.IsNothing = True [](start = 19, length = 37)

I don't think this does short-circuiting on Null in VB, consider doing explicit ```IsNot Nothing``` check, followed by IsNothing check

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Compilation/VisualBasicCompilation.vb:
```diff
@@ -1827,8 +1828,8 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
 
-            If result.IsReference AndAlso source.ConstantValue.HasValue AndAlso source.ConstantValue.Value Is Nothing Then
-                constantValue = source.ConstantValue
+            If result.IsReference AndAlso sourceConstantValue?.IsNothing = True Then
```
>sourceConstantValue?.IsNothing = True [](start = 42, length = 37)

Consider doing the same here

AlekseyTs commented:
>                     else if (!(operation.StepValue.GetConstantValue() is { IsBad: false }) &&

Is the new condition going to produce the same value when ```GetConstantValue()``` returns null

---
Refers to: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:4508 in a807e98. [](commit_id = a807e98b522bff02768b0975016003d2e95bcd35, deletion_comment = False)

AlekseyTs commented:
>                         if (value.Discriminator != ConstantValueTypeDiscriminator.Bad)

Given the updated if condition, can this ever be false?

---
Refers to: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:4682 in a807e98. [](commit_id = a807e98b522bff02768b0975016003d2e95bcd35, deletion_comment = False)

333fred commented:
>                     else if (!(operation.StepValue.GetConstantValue() is { IsBad: false }) &&

Yes. If the constant value is `null` or bad, we will take this path. There are tests that cover this.

---
In reply to: [649739664](https://github.com/dotnet/roslyn/pull/45127#issuecomment-649739664) [](ancestors = 649739664)

---
Refers to: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:4508 in a807e98. [](commit_id = a807e98b522bff02768b0975016003d2e95bcd35, deletion_comment = False)

333fred commented:
>                         if (value.Discriminator != ConstantValueTypeDiscriminator.Bad)

> Given the updated if condition, can this ever be false?

I don't believe so, and I believe that it never could have been in the first place. We never exposed bad `ConstantValue`s from `IOperation in the first place, so I'll remove this and make sure tests pass.

---
In reply to: [649742414](https://github.com/dotnet/roslyn/pull/45127#issuecomment-649742414) [](ancestors = 649742414)

---
Refers to: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:4682 in a807e98. [](commit_id = a807e98b522bff02768b0975016003d2e95bcd35, deletion_comment = False)

mjsabby commented:
I came here after my profiling an analyzer that does ConstantValue checks and is pretty slow because of it. Is there a way to determine if an expression is Constant without materializing the value (from an analyzer)?

333fred commented:
@mjsabby not currently, but that would be a reasonable API request IMO. Can you please file a new issue? It would be helpful to see your profiling results as well.

mjsabby commented:
@333fred Will do.

Analyzer code:

```csharp
var binaryExpressions = node.DescendantNodesAndSelf().OfType<BinaryExpressionSyntax>().Reverse();

foreach (var binaryExpression in binaryExpressions)
{
    var operation = semanticModel.GetOperation(binaryExpression);
    if (operation != null && operation.ConstantValue.HasValue) // allocating a ton
    {
        continue;
    }
}
```

The input looks like this:

```csharp
".024 0.463,-0.674 -0.188,-0.672 0.645,-0.035 -0.198,-0.652 1.767,-0.998 -0.144,-" +
"0.761 -1.874,-2.261 -2.103,-1.453 0.231,-2.973 0,0 1.027,0.489 -0.29,-0.807 0.70" +
"5,-1 -1.518,-1.259 0.102,-0.917 -0.484,-0.132 -0.139,-0.624 0.466,-0.913 0.572,0" +
".241 0.526,-0.701 0.62,-0.019 0.33,-1.167 0.414,0.27 0.566,-0.736 1.256,0.263 0." +
"013,0.512 0.368,0.198 0.068,-0.377 0.42,0.559 0.896,0.13 0.051,-0.54 -0.776,-0.7" +
"57 -0.308,-1.159 0.696,-1.06 -0.392,-0.092 0.272,-1.242 -1.834,-0.788 -0.599,-1." +
"079 -0.71,0.036 0.252,-1.042 -0.568,-0.319 0.576,-0.536 -1.843,-0.456 -0.399,-0." +
"941 0.548,-0.54 0.255,-2.525 1.482,-1.078 0.381,0.511 0.142,-0.359 0.633,0.072 -" +
"0.085,-1.638 1.026,0.258 0.51,-0.854 0.678,0.046 1.563,-1.789 1.461,-0.67 1.625," +
"-1.678 -0.509,-0.633 -0.627,0.135 -0.15,-0.432 0.784,-0.646 -0.814,-0.59 -0.058," +
"-0.677 -0.68,0.099 -0.309,-0.432 1.138,-0.972 -0.045,-0.457 -1.318,0.34 -0.479,0" +
".614 -0.822,-0.573 0,0 0.002,-0.009 0,0 -0.017,-0.079 0,0 1.206,-2.256 0.575,-3." +
"149 -0.302,-2.188 -0.778,-0.774 0.203,-0.515 -0.536,-0.694 0.303,-0.821 -0.889,-" +
"0.438 -0.325,-0.726 -0.82,-0.029 -0.347,-1.194 -0.618,-0.417 0.582,-1.532 -0.11," +
```
